### DISCUSSION TITLE
connect to external server (not localhost) then client.get_status() = 1 and i cant get a connection!

### DISCUSSION THREAD

**Discussion by felix7700** [Marzin-bot/PostgreSQLClient#31] [Q&A] [answered]
When I try to connect to an external psql server (not localhost) I always get client.get_status () = 1 and I can't connect!
I need help please!

**Comment by KuwazyStudio** [comment]
Your server may not be accessible from the outside. try to connect to your server using another postgres client to make sure the server is accessible.

**Reply by felix7700** [reply]
Many thanks for the answer! I've already checked that. I can successfully switch on via bash via psql command. but not with godot.
(I can easily access the same database via godot if it is located on the local psql server configured in exactly the same way, even with godot.)

**Reply by felix7700** [reply]
psql via bash works perfect.

/etc/postgresql/12/main/pg_hba.conf  
host    all             all             0.0.0.0/0            md5

/etc/postgresql/12/main/postgresql.conf
listen_addresses = '*'  

sudo psql -h 192.168.1.140 -d postgres -U postgres
Password for user postgres: 
psql (12.8 (Ubuntu 12.8-0ubuntu0.20.04.1))
SSL connection (protocol: TLSv1.3, cipher: TLS_AES_256_GCM_SHA384, bits: 256, compression: off)
Type "help" for help.

postgres=#

**Comment by felix7700** [comment]
![Bildschirmfoto vom 2021-11-01 20-50-05](https://user-images.githubusercontent.com/51541572/139735755-75b0939f-2267-470d-83e2-e6f7e66b8161.png)
![Bildschirmfoto vom 2021-11-01 20-59-54](https://user-images.githubusercontent.com/51541572/139735760-89393005-bd34-4401-87f0-3c4008f06c6d.png)
on local host it works perfect, and in the last line in the consol i get my table-items and client.get_status() = 2

**Reply by felix7700** [reply]
But if I change the ip to access my external psql server (same psql server settings, same user, same password) I don't get a connection.

**Comment by felix7700** [comment]
![external_no_connection](https://user-images.githubusercontent.com/51541572/139736843-a0a010eb-0064-453b-ab3d-3736e2dcf547.png)

**Comment by KuwazyStudio** [comment]
do you use ssl when logging in?

**Comment by KuwazyStudio** [comment]
in the ssl parameter of the connect_to_host() method.

**Reply by felix7700** [reply]
yes it seems like ssl is being used, ssl is set to true in the script
func connect_to_host(url: String, ssl := true, connect_timeout := 30) -> int:

**Reply by KuwazyStudio** [reply]
pass the parameter to false to see.

**Comment by felix7700** [comment]
![extern_no_connection_ssl_off](https://user-images.githubusercontent.com/51541572/139739583-3d844d20-4c88-4862-a393-a6f27d6989fb.png)

exactly the same problem

**Comment by KuwazyStudio** [comment]
It's strange to me it works. I am using the new version of PostgreSQLClient for godot 3.4. Maybe you should test it. In the meantime I will investigate the problem.

**Comment by KuwazyStudio** [comment]
no doubt a problem with the tcp socket.

**Comment by KuwazyStudio** [comment]
The version for godot 3.4 has not yet been released in the assetLib. You need to download the main branch of this reference.

**Comment by KuwazyStudio** [comment]
Please report debug errors and notices related to PostgreSQLClient to me. Thank you for reporting your issue. I will write here if there is anything new about the problem.

**Comment by KuwazyStudio** [comment]
Feel free to use breakpoints in the code to see which line exactly is blocking.

**Comment by felix7700** [comment]
the problem seems to be in the function:
func connect_to_host(url: String, ssl := false, connect_timeout := 30) -> int:
in the line:
if error == OK and client.is_connected_to_host() and client.get_status() == StreamPeerTCP.STATUS_CONNECTED:

the client.get_status() value is always 1 (connecting) when i try to connect to a external psql-server, but the value has to be 2 (connected)

**Comment by KuwazyStudio** [comment]
That's good I see where the problem comes from! It depends on the connection time to the server. In fact I am not checking that the connection has been established in an infinite loop. The problem occurs when the server takes a long time to establish the connection. This also explains why it works with "localhost". My server was quick to establish the connection, so I didn't see the problem.

**Comment by KuwazyStudio** [comment, marked as answer]
I updated the main branch. Download it and test it. The problem should be resolved. Note that the main branch is not finished so I will invite you to re-download it when I publish the version to the assetLib.

**Comment by felix7700** [comment]
now i get a error message
![error](https://user-images.githubusercontent.com/51541572/139748860-f9251d03-b4f4-4541-bad0-01a1b131d4d4.png)

**Reply by KuwazyStudio** [reply]
If you have this error it means that the other was successfully resolved. now look at this problem: https://github.com/Marzin-bot/PostgreSQLClient/issues/30

**Comment by felix7700** [comment]
Now i am using godot v3.4 beta1 and it works! Thanks very much!

(In the function connect_to_host (): client.get_status () is still 1, I don't know why now, shouldn't the value be 2? But hey, it works! ;-)

**Reply by KuwazyStudio** [reply]
she should. I will correct that. Please help me solve this problem!

**Reply by KuwazyStudio** [reply]
Having to add a While in connect_to_host () makes it blocking. I'll have to update the script to make it non-blocking. Do try to follow the project to be aware of updates.

**Comment by KuwazyStudio** [comment]
The state is 2 when the connection established signal () is output.

**Comment by felix7700** [comment]
Now I have found that with the current PostgreSQLClient.gd version I can access an external server, but no longer a localhost, which would also be important for the test environment for developing my software.

**Comment by KuwazyStudio** [comment]
Hello, I will investigate this problem right away. Anyway all the tests done before release worked .... Note that I'm only 18 and started programming in GDScript 3 years ago so I'm sorry if you encounter problems in my projects. 😭

**Comment by felix7700** [comment]
I changed the following and now the connection to the local host works.


var counter := 0
while not (error == OK and client.is_connected_to_host() and client.get_status() == StreamPeerTCP.STATUS_CONNECTED) or counter == 0:
counter += 1

**Reply by felix7700** [reply]
(Sorry that I found another bug. I know you work on this project in your spare time. Nevertheless, I would like to thank you for working on this project! Thanks very much! You're doing a good job, keep it up!)

**Comment by KuwazyStudio** [comment]
You don't have to apologize. On the contrary, you help the project to become better. I updated, please test PostgreSQLClient locally and remotely. 😄

**Comment by KuwazyStudio** [comment]
Please feel free to report any bugs you may encounter. :+1:
